src/applications/lgy/coe/components/CoeDocumentUpload.jsx

```diff
@@ -1,7 +1,129 @@
+
+const initialState = {
+  documentType: DOCUMENT_TYPES[0],
+  documentDescription: {
+    dirty: false,
+    value: '',
+  },
+  errorMessage: null,
+  files: [],
+};
+
+const {
+  DOC_TYPE,
+  DOC_DESC,
+  FILE_UPLOAD_SUCCESS,
+  FILE_UPLOAD_FAIL,
+  FILE_UPLOAD_PENDING,
+  FORM_SUBMIT_FAIL,
+  DELETE_FILE,
+} = ACTIONS;
+
+const scrollToError = () => {
+  const errors = document.querySelectorAll('.usa-input-error');
+  if (errors.length) {
+    const errorPosition = getTopPosition(errors[0]);
+    const options = getScrollOptions({ offset: -25 });
+    Scroll.animateScroll.scrollTo(errorPosition, options);
+    errors[0].querySelectorAll('label')[1].focus();
+  }
+};
+
+const reducer = (state, action) => {
+  switch (action.type) {
+    case DOC_TYPE:
+      return { ...state, documentType: action.documentType };
+    case DOC_DESC:
+      return { ...state, documentDescription: action.documentDescription };
+    case FILE_UPLOAD_SUCCESS:
+      return {
+        ...state,
+        files: [...state.files, action.file],
+        documentType: DOCUMENT_TYPES[0],
+        documentDescription: {
+          dirty: false,
+          value: '',
+        },
+        errorMessage: null,
+      };
+    case FILE_UPLOAD_PENDING:
+      return { ...state, errorMessage: null };
+    case FORM_SUBMIT_FAIL:
+    case FILE_UPLOAD_FAIL:
+      return { ...state, errorMessage: action.errorMessage };
+    case DELETE_FILE:
+      return { ...state, files: action.files };
+    default:
+      return state;
+  }
+};
 
 export const CoeDocumentUpload = () => {
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  const { documentType, documentDescription, errorMessage, files } = state;
+
+  const errorMsgClass = errorMessage ? 'vads-u-padding-left--1p5' : null;
+  const disabledOnEmptyDescClass =
+    documentType === 'Other' && !isNotBlank(documentDescription.value)
+      ? 'file-input-disabled'
+      : null;
+
+  const onSelectChange = e => {
+    dispatch({ type: DOC_TYPE, documentType: e?.value });
+  };
+
+  const onTextInputValueChange = e => {
+    dispatch({
+      type: DOC_DESC,
+      documentDescription: { dirty: true, value: e?.value },
+    });
+  };
+
+  const onUploadFile = async uploadedFiles => {
+    dispatch({ type: FILE_UPLOAD_PENDING });
+    if (!isValidFileType(uploadedFiles[0])) {
+      dispatch({
+        type: FILE_UPLOAD_FAIL,
+        errorMessage:
+          'Please choose a file from one of the accepted file types.',
+      });
+      return;
+    }
+    const file = uploadedFiles[0];
+    file.documentType = documentType;
+    if (documentDescription.value !== '') {
+      file.documentDescription = documentDescription.value;
+    }
+    dispatch({ type: FILE_UPLOAD_SUCCESS, file });
+  };
+
+  const onDeleteFile = idx => {
+    const newFiles = state.files.filter((file, index) => index !== idx);
+    dispatch({ type: DELETE_FILE, files: newFiles });
+  };
+
+  const onSubmit = () => {
```

Comment:
Work to stand up an endpoint in vets-api where these files can live hasn't begun yet. A future pull request will include the additional work of handling this http request.